docs/source/column-metadata.rst

```diff
@@ -0,0 +1,185 @@
+
+Implementation
+==============
+
+Regardless of how these properties change computation, ``blaze`` needs to be able to tell other systems about metadata where possible. This is one of the main benefits to using a relational database, namely, the ability to relate information.
+
+Sorted
+------
+
+TODO
+
+Unique and Grouped Attributes
+-----------------------------
+
+Categorical Types
+^^^^^^^^^^^^^^^^^
+Categorical types can cover the cases of unique and grouped. We implement this in the ``datashape`` library. It has two relevant properties:
+
+1. A list of unique sorted categories.
+2. A list of the indices of the categories from item 1 as unsigned integers of the minimum type necessary to support the number of unique categories.
+
+Primary and Foreign Keys
+------------------------
+
+It is important to note that we want to keep foreign key metadata separate from type information. Adding foreign key metadata to a blaze expression will thus require an extra step.
+
+   .. code-block:: python
+
+      >>> t = symbol('t', """var * {
+      ...     id: int64,
+      ...     name: categorical[type=string, categories=["Alice", "Bob", "Joel"]],
+      ...     amount: float64
+      ... }""")
+      >>> t.primary_key = t.id
+      >>> s = symbol('s', "var * {t_id: int64, height: float64}")
+      >>> s.relations = {s.t_id: t.id}
+
+
+Alternatively we could allow ``primary_key`` and ``relations`` keyword arguments:
+
+   .. code-block:: python
+
+      >>> t = symbol('t', """var * {
+      ...     id: int64,
+      ...     name: categorical[type=string, categories=["Alice", "Bob", "Joel"]],
+      ...     amount: float64
+      ... }""", primary_key='id')
+      >>> s = symbol('s', "var * {t_id: int64, height: float64}", relations={'t_id': t.id})
+
+
+A more general approach might be to pass around a dictionary of arbitrary metadata. This has the benefit of not tying the notion of relations and keys to blaze's ``Symbol`` objects, which are meant to be able to describe many different kinds of data other than tables.
+
+
+   .. code-block:: python
+
+      >>> t = symbol('t', dshape=..., meta={'keys': {'primary': ['id']}})
+      >>> s = symbol('s', dshape=..., meta={'keys': {'foreign': {'t_id': t.id}}})
+
+This approach also has the benefit of being low tech. We're passing around core Python data structures (``list`` and ``dict``) that fully describe primary and foreign key relationships.
+
+
+Validation can also take place at construction time. A few things can be validated:
+
+* The ``keys`` dictionary key is only valid on ``Record`` dshapes
+* The primary key is actually in the datashape
+* The primary key is a valid primary key type (this would likely be limited to integers at first)
+* Foreign keys all reference primary keys
+
+
+In any case we want to be able to access implicit join functionality:
+
+   .. code-block:: python
+
+      >>> compute(s.t_id.amount, {s: <sdata>, t: <tdata>})
```

Comment:
It might be worth looking at SQLAlchemy's ORM to see how they spell these things.